src/celeritas/phys/PhysicsStepUtils.hh

```diff
@@ -341,11 +341,22 @@ select_discrete_interaction(const ParticleTrackView& particle,
         }
     }
 
-    // Select the model and return; See doc above for details.
+    // Find the model that applies at the particle energy
     auto find_model = physics.make_model_finder(ppid);
-    auto model_id   = find_model(particle.energy());
-    CELER_ENSURE(model_id);
-    return physics.model_to_action(model_id);
+    auto pmid       = find_model(particle.energy());
+
+    ElementComponentId elcomp_id{0};
+    if (auto table_id = physics.value_table(pmid))
```

Comment:
Your current implementation works since we *always* sample the element in the on-the-fly model kernels, but it's in error if we need to check `!sampled` in case of a kernel relaunch. So it would be safer perhaps to pass a material view here and add:
```suggestion
    ElementComponentId elcomp_id{};
    if (mat.num_elements() == 1)
    {
        elcomp_id = ElementComponentId{0};
    }
    else if (auto table_id = physics.value_table(pmid))
```

Reply:
Good point, done.